Support opensearch-ruby 4.0

issei-m opened the pull request:
### Description

As while discussions in this [PR](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/46), we figured out the upcoming `opensearch-ruby` broken compatibility a bit but makes our library completely being no longer working.

So this PR resolves the issue.

### Issues Resolved

n/a

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock reviewed (commented):
First, thanks for picking this up!

So, this replaces the transport in the subclassed client? I think we should not rely on assigning `@transport`. I'll merge this once we're done with the 3.x line as it makes things work.

As a next step however I think we should revisit this approach. My first idea was to expose `transport` as a client option in the 4.0 opensearch-ruby and getting rid of `Sigv4Client` altogether. Users would instantiate a `Sigv4Transport`.

But IMO this is wrong too. A sigv4 signer is about request signing. I think users want to pass it as an `auth:` parameter, and the opensearch-ruby library should support basic auth, sigv4, and others this way. WDYT?

issei-m commented on lib/opensearch-aws-sigv4.rb:
```diff
@@ -44,68 +113,12 @@ class Sigv4Client < ::OpenSearch::Client
+        @transport = OpenSearch::Aws::Transport::Sigv4Client.new(transport_args, sigv4_signer, options[:sigv4_debug],
+                                                                 &block)
       end
+      # rubocop:enable Lint/MissingSuper
```
Yes. I don't want the superclass to meaninglessly instantiate a hard-coded `transport`, that will be overwritten soon after.

dblock replied:
Makes sense, this "interface" is borked, see my comment above for what I think we should do here. LMK what you think!

nhtruong replied:
@dblock does adding sigv4 as an auth option in `opensearch-ruby` basically moves this feature back into `opensearch-ruby`? I couldn't find the discussion where we decided to move sigv4 out the core library but I remember we extracted sigv4 out of `opensearch-ruby` because sigv4 is a propriety auth method owned by Amazon, and as an opensource project we do not want any favoritism by including Amazon's sigv4 and its dependencies in `opensearch-ruby` gem.

nhtruong replied:
Gotcha. I've created a separated [issue](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/issues/50) for that :)

issei-m commented:
My initial idea was quite similar, and what I was actually trying to do was something like this:

```ruby
signer = Aws::Sigv4::Signer.new(...)

transport = OpenSearch::Transport::HTTP::Faraday.new(
  url: 'http://localhost:9200',
  options: {
    headers: { 'Content-Type' => 'application/json' },
    transport_options: {
      ssl: { verify: false }
    }
  }
)

sigv4_signing_transport_decorator = OpenSearch::Aws::Transport::Transport::Sigv4SigningTransportDecorator.new(transport, signer)

client = OpenSearch::Client.new({ transport: sigv4_signing_transport_decorator })
```

This decorator simply receives perform_request and delegates the request to transport using the signed headers. In fact, this approach looks clean. However, the decorated OpenSearch::Transport::Transport::Base actually has many public methods, and I'm a bit concerned that if a decorator is created too easily, it might not behave as intended by the base library. In reality, it doesn't seem like perform_request is being called from a higher-level context, but it might be worth reviewing this a bit more.

dblock commented:
@issei-m It's quite unusual to subclass a transport for the purposes of signing requests, try to change the implementation to support callbacks that allow for adding headers instead?

nhtruong commented:
The solution that @issei-m has here makes sense! It makes this lib work with `opensearch-ruby 4` the same way that it did with `opensearch-ruby 3`:

- The `opensearch-ruby` prior to 4.x had a wacky way of invoking the `transport.perform_request` method via  the `client.perform_request` method. Version 4.0 refactored this part of the library and removed the middle-man methods, including `client.perform_request` (Check the **How the perform_request method is invoked** section in this [PR](https://github.com/opensearch-project/opensearch-ruby/pull/261)
- This sigv4 library intercepts `client.perform_request` (via sub-classing the `OpenSearch::Client` class to override this method) to inject the sigv4 signature. However, since `client.perform_request` has been removed in 4.x and all API methods now call the `transport.perform_request` directly, @issei-m's solution is overriding the transport's method instead, and without introducing breaking changes for the current sigv4 users.

Instead of us asking the user every bit of auth info to perform a sigv4 request, we delegate this to the official Sigv4 gem, as the user has to provide an instance of the `Aws::Sigv4::Signer` class when instantiating the Sigv4 client. This insulate us from changes in Sigv4 (as it will be handled by the AWS Sigv4 gem), and allows the user to use any version of Sigv4.

nhtruong commented:
The decorator approach described [here](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/47#issuecomment-2676921084) is a bit too complicated to use, IMO. It requires the user to know the implementation details of the transport layer and create an instance of `OpenSearch::Transport::HTTP::Faraday`

nhtruong commented:
This looks good to me. Just need to rebase and update the change log.

I don't think this will introduce breaking changes and we can release this as a minor version? What do yall thing?

dblock commented:
> I don't think this will introduce breaking changes and we can release this as a minor version? What do yall thing?

I would add CI with opensearch-ruby 3.x and 4.x to be sure.

issei-m commented:
@nhtruong Thanks for your review and approval.  

You mean we can go with this approach for both opensearch-ruby 3.x and 4.x?  

If so, I need to reintroduce the `ignore` param removal I incorporated in my previous PR as I've removed it while porting the signature-making method to the transport layer in [this PR](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/47/files#diff-eda7c3168a0d0c0ac8e4fcd9235f48492fc2fa92530236b11e0e8e592e60e203R56), since I originally thought this would be shipped for only 4.x line.

nhtruong commented:
@issei-m the `ignore` flag has always been a feature of the Ruby client. Its implementation is confusing due to how it's passed to the method, and doing it differently will introduce a breaking change. Since we the next major release is still in beta, we should consider re-implementing this feature. What do you think?

cc @dblock

issei-m commented:
@nhtruong IMPO, mixing the client options with the hash that holds the OpenSearch server options is confusing. If breaking changes are allowed, it might be worth considering a different approach (such as nesting them under `client_options`). 

Also, since there is already `allow_no_indices`, maybe this option itself isn’t necessary? (Errors other than 404 should generally be handled by the client caller.)

nhtruong commented:
I don't think `allow_no_indices` is the same as the `ignore` flag in the Ruby client. While most APIs return a `200 OK` when you try to delete a resource that has already been deleted (i.e. the delete action is idempotent), OpenSeach API returns a `404` instead. `ignore: 404` in the Ruby client attempts to simulate this convention while interacting with an OpenSearch cluster.

I'll come up with a PR for OpenSearch Ruby Beta.2 soon that will move this feature into the client_options while instantiating the client, instead of the user passing this flag in every API method call. Then with version 4, Sigv4 and other wrappers wont have to worry about excluding the `ignore` params anymore.

issei-m commented:
@nhtruong Ah, it looks like I misunderstood `allow_no_indices`.  
btw all CI failures have been resolved now. I've kept the `ignore` param removal since I'm not sure how the beta library will evolve. If it turns out to be unnecessary, please remove it.

nhtruong commented:
Yeah let's hold on to this PR for now while I work on reworking this feature for the beta client.

nhtruong commented:
@issei-m OpenSearch Ruby [4.0.0.pre.beta.2](https://rubygems.org/gems/opensearch-ruby/versions/4.0.0.pre.beta.2) has been released. It has moved the idempotent delete to the [client level](https://github.com/opensearch-project/opensearch-ruby/blob/main/guides/idempotent_delete.md). That means when ignore is passed to an API action, it will be treated as a query param, and this Sigv4 gem does not have exclude that param when performing the signing anymore.

We just need to bump the [dependency's version](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/47/files#diff-5bbb559f68aa5b9eba15f1eb444eba67e88861379c522b22bc77cb88e73a66eaR53) to `beta.2` and remove the 2 `params.delete(:ignore)` [lines](https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/47/files#diff-eda7c3168a0d0c0ac8e4fcd9235f48492fc2fa92530236b11e0e8e592e60e203R57-R58).

issei-m commented:
Integration test has failed, I'm going to check them out.

issei-m commented:
I’ve upgraded `rubocop-rake` and `rubocop-rspec` because the previous combination didn’t seem to be compatible with the latest `rubocop`. (It might be better to apply it to a lower version branch as well.)

dblock commented on lib/opensearch-aws-sigv4.rb:
```diff
@@ -17,6 +17,74 @@
+      class Sigv4Client < ::OpenSearch::Transport::Client
+        attr_accessor :sigv4_signer
+
+        def initialize(transport_args, sigv4_signer, sigv4_debug, &block)
```
It's really uncommon to have a constructor like this. It is difficult to evolve, and does not match the superclass.

Next, the `sigv4_debug` parameter is problematic because it directs security-related information to logger in an unconventional way. Get rid of it and always `log` Sigv4 information at debug level. This way when auth fails using this client the user can turn on debug level logging and narrow things down without modifying the code.

So, the constructor should match the one of `OpenSearch::Transport::Client`

```ruby
def initialize(arguments = {}, &block)
   arguments = arguments.dup
   signer = arguments.delete(:signer) 
   raise ... unless signer
   super arguments, block
end
```

dblock commented on lib/opensearch-aws-sigv4.rb:
```diff
@@ -17,6 +17,74 @@
+          logger.debug("#{title.upcase}:\n\e[36m#{message}\e[0m")
+        end
+
+        def logger
```
Definitely an antipattern, just use the logger provided by the caller, don't instantiate one here.

dblock commented on lib/opensearch-aws-sigv4.rb:
```diff
@@ -17,6 +17,74 @@
+
+          log('string to sign', signature.string_to_sign)
+          log('canonical request', signature.canonical_request)
+          log('signature headers', signature.headers)
```
This class isn't the right place to log request headers, I bet it if you turn on debug logging the transport will do it. I would log the size of data being signed (not the data, it could contain PII!) and the size of the resulting signature. The signature itself will get logged by the transport when sent to the server.

dblock commented on lib/opensearch-aws-sigv4.rb:
```diff
@@ -36,6 +104,7 @@ module Aws
     #   puts client.cat.health
     #
     # @attr [Aws::Sigv4::Signer] sigv4_signer Signer used to sign every request
+    # rubocop:disable Lint/MissingSuper
```
This is not a cop you want to disable. Even if we subclass the client, call `super`. The signer should be extracted out of the parameters as I mentioned above.

dblock reviewed (changes requested):
I think this is the time when we can fix the constructor for the transport signer at least. 

Alternatively we can merge this as is, but I definitely would not want a major release without fixing this and #50.

nhtruong commented:
@issei-m feel free to create another PR or modify this one for the misc changes in this PR. (Sorry I missed the notification on your response)

nhtruong commented:
@dblock This [PR](https://github.com/opensearch-project/opensearch-ruby/pull/281) on the OS Ruby will allow use to turn this gem into an auth callback with simpler implementation of sigv4 signing.

issei-m commented:
@nhtruong I'm working on another branch

issei-m commented:
Closed this PR as I'm reworking it here https://github.com/opensearch-project/opensearch-ruby-aws-sigv4/pull/62
